src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Rules/CannotChangeGenericConstraints.cs

```diff
@@ -0,0 +1,167 @@
+            var rightTypeParameters = rightType.TypeParameters;
+
+            // can remove constraints on sealed classes since no code should observe broader set of type parameters
+            bool permitConstraintRemoval = !_settings.StrictMode && leftType.IsSealed;
+
+            CompareTypeParameters(leftTypeParameters, rightTypeParameters, left, permitConstraintRemoval, leftMetadata, rightMetadata, differences);
+        }
+
+        private void RunOnMemberSymbol(
+            ISymbol? left,
+            ISymbol? right,
+            ITypeSymbol leftContainingType,
+            ITypeSymbol rightContainingType,
+            MetadataInformation leftMetadata,
+            MetadataInformation rightMetadata,
+            IList<CompatDifference> differences)
+        {
+            if (left is not IMethodSymbol leftMethod || right is not IMethodSymbol rightMethod)
+            {
+                return;
+            }
+
+            var leftTypeParameters = leftMethod.TypeParameters;
+            var rightTypeParameters = rightMethod.TypeParameters;
+
+            bool permitConstraintRemoval = !_settings.StrictMode && !leftMethod.IsVirtual;
+
+            CompareTypeParameters(leftTypeParameters, rightTypeParameters, left, permitConstraintRemoval, leftMetadata, rightMetadata, differences);
+        }
+
+        private void CompareTypeParameters(
+            ImmutableArray<ITypeParameterSymbol> leftTypeParameters,
+            ImmutableArray<ITypeParameterSymbol> rightTypeParameters,
+            ISymbol left,
+            bool permitConstraintRemoval,
+            MetadataInformation leftMetadata,
+            MetadataInformation rightMetadata,
+            IList<CompatDifference> differences)
+        {
+
+            Debug.Assert(leftTypeParameters.Length == rightTypeParameters.Length);
+            for (int i = 0; i < leftTypeParameters.Length; i++)
+            {
+                ITypeParameterSymbol leftTypeParam = leftTypeParameters[i];
+                ITypeParameterSymbol rightTypeParam = rightTypeParameters[i];
+
+                // Currently our symbol formatting for comparison includes generic parameter names rather than just arity
+                // https://github.com/dotnet/roslyn/issues/73051
+                // As a result this will only be called with matching parameters.
+                // When parameters don't match the symbol will be treated as different because it's key repesentation will differ.
+                // See SymbolEqualityComparer.GetKey https://github.com/dotnet/sdk/blob/5c8c0a75bba35c1fd0db879eb821c8635c142f79/src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Comparing/SymbolEqualityComparer.cs#L22-L41
+                Debug.Assert(leftTypeParam.Name.Equals(rightTypeParam.Name));
+
+                List<string> addedConstraints = new();
+                List<string> removedConstraints = new();
+
+                CompareBoolConstraint(typeParam => typeParam.HasConstructorConstraint, "new()");
```

Comment:
Do we really need to hardcode the constraint syntax? Isn't there an API that returns the values?

Reply:
These are just used for log messages / target strings.  They could be nonsense if we wanted them to be.  I'll double check if roslyn defines any constants publicly.

Reply:
I couldn't find any public API in roslyn that exposes these. Internally they are listed as literal strings, or as keywords, but no easily consumable public API exposes single constraint syntax for these.  I'm not too worried about this single instance of checked in syntax since it can never change and matches the public docs for constraints.